src/vmemcache_heap.c

```diff
@@ -198,59 +226,105 @@ vmcache_heap_destroy(struct heap *heap)
 {
 	LOG(3, "heap %p", heap);
 
-	VEC_DELETE(&heap->entries);
 	util_mutex_destroy(&heap->lock);
 	Free(heap);
 }
 
+/*
+ * vmcache_free_first_extent -- (internal) free the first extent
+ */
+static int
+vmcache_free_first_extent(struct heap *heap, struct extent *ext)
+{
+	ASSERTne(ext->ptr, NULL);
+
+	struct header *first_header = HEADER(ext->ptr);
+	ASSERTeq(first_header->prev, NULL);
+	ASSERTne(first_header->next, NULL);
+
+	void *second_extent = first_header->next;
+	struct header *second_header = HEADER(second_extent);
+	ASSERTeq(second_header->prev, ext->ptr);
+
+	/* remove the first extent from the list */
+	second_header->prev = NULL;
+
+	/* free the first extent */
+	struct heap_entry he;
+	vmcache_heap_merge(heap, ext, &he);
+	vmcache_insert_heap_entry(heap, &he, &heap->last_extent, IS_FREE);
+
+	__sync_fetch_and_sub(&heap->size_used, ext->size);
+
+	return 0;
+}
+
 /*
  * vmcache_alloc -- allocate memory (take it from the queue)
  *
- * This function returns the number of allocated bytes if successful,
- * otherwise -1 is returned.
+ * It returns the number of allocated bytes if successful, otherwise -1.
+ * The last extent of doubly-linked list of allocated extents is returned
+ * in 'last_extent'.
+ * 'first_extent' has to be zeroed in the beginning of a new allocation
+ * (e.g. when *last_extent == NULL).
  */
 ssize_t
-vmcache_alloc(struct heap *heap, size_t size, struct extent_vec *vec)
+vmcache_alloc(struct heap *heap, size_t size, void **last_extent,
```

Comment:
I think this patch would be more readable with the extent lists beginning at the front ;)

Reply:
Why? I am not sure about it.